### ISSUE TITLE
Cordon node based on expire after value

### ISSUE THREAD

**Issue by ugurgural** [aws/karpenter-provider-aws#5664]
Karpenter version: v0.33.1
Kubernetes version: 1.28

Hi, I have long-running ml jobs on CI and both CI agents and experiment pods are managed by karpenter. For cost effective approach I am using different node pool for CI agents, so they can schedule on cheaper EC2 instances. My problem is I need to expire the nodes containing CI agent pods after some time(usually around a week), so they can always run on latest updated AMIs. 

I tried `expireAfter` flag, which kind of works but it entirely drains the node once the value in that flag mets the criteria. Tried do-not-evict tag on CI agent pods, but that only works before node goes into deletion phase. If there is a new agent pod coming to that node, it does not know that this node will be terminated, hence it gets deleted in the process, causing downtime.

So long story short, I need a mechanism that after X amount of time, it cordons the node, waits for the workload on the node to be end, don't get anything new to that node and terminate the node at the end. How can I tackle this ? For example this project looks like doing what I need: https://github.com/XenitAB/node-ttl

Thank you for the answers.

**Comment by jonathan-innis** [comment]
> it cordons the node, waits for the workload on the node to be end, don't get anything new to that node and terminate the node at the end

I'd say take a look at this issue: https://github.com/kubernetes-sigs/karpenter/issues/622. We are planning on adding support for a taint to get added to the node as soon as we meet a disruption condition. The debate right now is basically between whether this should be a `PreferNoSchedule` or an outright `NoSchedule`. In either case, this would push your application pods away from scheduling on these nodes so that you would avoid the issue that you are running into here.

**Comment by jonathan-innis** [comment]
Closing as a duplicate of https://github.com/kubernetes-sigs/karpenter/issues/622
